aten/src/THS/generic/THSTensor.cpp

```diff
@@ -283,6 +304,27 @@ THSTensor *THSTensor_(newTranspose)(THSTensor *self, int d1, int d2) {
  * reshaping methods
  ******************************************************************************/
 
+/*** Helper methods ***/
+THLongTensor *THSTensor_(toCSR)(int64_t const *indices, int64_t dim, int64_t nnz) {
+  int64_t h, i, hp0, hp1;
+  THLongTensor *csr = THLongTensor_newWithSize1d(dim + 1);
+  THLongTensor_zero(csr);
+
+  // Convert the sparse matrix to CSR format
+#pragma omp parallel for private(i, h, hp0, hp1) schedule(static) if (nnz > 10000)
+  for (i=0; i<nnz; i++) {
+    hp0 = indices[i];
+    hp1 = (i+1 == nnz) ?  dim : indices[i+1];
+    if (hp0 != hp1) for (h = hp0; h < hp1; h++) {
+      THTensor_fastSet1d(csr, h+1, i+1);
+    }
+  }
+  return csr;
```

Comment:
This also looks as if it assumed that the indices are for a 1-d sparse tensor. You need to have a check somewhere

Reply:
I also don't quite understand what the code does, do you have any reference for it? Isn't CSR for a 1D tensor equivalent to COO?

Reply:
`toCSR` is not a new function. It was in `THSTensorMath.cpp` and`THCSTensorMath.cu` as an internal method for matrix multiplication, and it is moved to `THSTensor.cpp` and `THCSTensor.cu` in this pull request.

COO is represented by row indices and column indices and CSR is represented by row "pointers" and column indices. In order to convert a (coalesced and sorted) COO representation to CSR representation, we need to convert row indices into row pointers, while we don't need to do anything on column indices. 

`toCSR` is the function that converts row indices (1-D) into row pointers (1-D). For GPU version, it wraps [`cusparseXcoo2csr`](https://docs.nvidia.com/cuda/cusparse/index.html#cusparse-lt-t-gt-coo2csr). 
> This function converts the array containing the uncompressed row indices (corresponding to COO format) into an array of compressed row pointers (corresponding to CSR format).

The CPU version implements the equivalent operation in openMP. The input row indices MUST be sorted, or the result will be incorrect.

Reply:
This function gets a standard 2-d `indices` tensor viewed as a 1-d array of long when called (see [L561](https://github.com/pytorch/pytorch/pull/6225/files/36acd951e6c6b4d1fd76bb0fddd527b6b36ce4e4#diff-3aa2a0615d10b1a06bd582a0d1bbf61dR561)), and uses only the first `nnz` elements of it. I will NOT add a check for this one.

Reply:
We do need to check for it, because it will be incorrect for multidimensional arrays. Just push the viewing code inside this function (take in a raw `LongTensor` indices), and add error checking here.